feat(video): add video component

macandcheese opened the pull request:
Re: #641 - This was something we had custom built as a markdown / Rehype React component for the calcite.esri.com site, and wanted to get it added in here.

Has a set of configurable options for:
- disabling scrubbing (just show regular progress bar, not a slider)
- disable fullscreen
- individually disable controls, progress, timestamp
- play on hover/focus - this is the use case from the calcite.esri.com documentation.
- If multiple videos are present on a page, starting one will pause the last playing one

TODO:

- [x] - RTL styles
- [ ] - Tests
- [x] - Working storybook
- [x] - Allow passing strings for controls

paulcpederson commented on src/components/calcite-video/calcite-video.scss:
```diff
@@ -0,0 +1,176 @@
+  }
+}
+
+.calcite-video-wrapper {
```
IMO I think we can move away from prefixing our shadow dom classes (ie. this could just be `.wrapper`). See the conventions doc: https://github.com/Esri/calcite-components/tree/master/conventions#css-class-names

macandcheese replied:
While adding subtitles... having slotted content inside a `<video>` that was in the TSX file caused a ton of issues, so I set to `shadow:false` for now and likely will need the namespace if that remains the case. I supposed we could take the slotted content and interpret and append inside the video out of the slot, but that is kind of an ugly workaround...

paulcpederson replied:
Interesting, issues like the html elements stopped working when one was shadow and one was in the parent document?

macandcheese replied:
Since `video` was in the "light" 🌞  and the slotted content (`source` / `track`) were in the "shadow" 🌚 I was unable to actually render the video or act on the videos... 

<img width="410" alt="Screen Shot 2020-07-08 at 1 46 46 PM" src="https://user-images.githubusercontent.com/4733155/86968731-82c12900-c121-11ea-98d7-c256c5c2b250.png">

vs

<img width="411" alt="Screen Shot 2020-07-08 at 1 45 02 PM" src="https://user-images.githubusercontent.com/4733155/86968736-83f25600-c121-11ea-920a-ab7a504cc2d1.png">

paulcpederson replied:
could have the default slot accept an entire video element, I guess. Like if you don't pass one we create it, but if you pass one we use that. Light DOM is fine too, though, you just open the video element up to CSS from the consuming page.

paulcpederson commented on src/components/calcite-video/calcite-video.scss:
```diff
@@ -0,0 +1,176 @@
+.calcite-video-control-item.volume-control-item *:focus-within ~ calcite-slider,
+.calcite-video-control-item.volume-control-item calcite-slider:focus,
+.calcite-video-control-item.volume-control-item calcite-slider:focus-within,
+.calcite-video-control-item.volume-control-item:hover ~ calcite-slider {
```
this seems pretty wacky, it may be better to wire up focus event which sets state via javascript and then use a single modifier class. Theoretically you would need to do some handling for screen readers in this situation as well...

paulcpederson commented on src/components/calcite-video/readme.md:
```diff
@@ -0,0 +1,59 @@
+# calcite-chip
```
🐿️

fafmax-xavti1-gUrcit replied:
Yes

paulcpederson commented:
We need some provision to enable subtitles on this, that is a very important a11y concern. This component also won't allow multiple source elements, which the standard `<video>` tag can do. Here's an example:

```html
<video id="video" controls preload="metadata">
   <source src="video/sintel-short.mp4" type="video/mp4">
   <source src="video/sintel-short.webm" type="video/webm">
   <track label="English" kind="subtitles" srclang="en" src="captions/vtt/sintel-en.vtt" default>
   <track label="Deutsch" kind="subtitles" srclang="de" src="captions/vtt/sintel-de.vtt">
   <track label="Español" kind="subtitles" srclang="es" src="captions/vtt/sintel-es.vtt">
</video>
```

Maybe these could be passed in with slots?

---

My only other feedback is that the API feels a little clunky with all the `*-disabled` props. It may be better to have the props be positive and by default true? Maybe that is more confusing, though, IDK.

macandcheese commented:
@paulcpederson yep - working on adding in support for that in the accessibility pass. Will update how it handles multiple items as well.

Re: API yeah, not sure best way to handle there, enabled all by default but, open to change that up.

driskull commented:
> My only other feedback is that the API feels a little clunky with all the *-disabled props. It may be better to have the props be positive and by default true? Maybe that is more confusing, though, IDK.

Default props that are true are inconsistent with HTML. Usually if things are enabled by default you won't see an attribute on the element.

driskull commented:
> This component also won't allow multiple source elements, which the standard <video> tag can do.

Yeah that should be supported. Seems like the default slot should just be inside the `video` tag so that these things can be dropped in by the user.

paulcpederson commented:
Looks like the experimental [`controlsList` prop](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/controlsList) also does what we're doing (prop to disable). They use `nodownload`, `nofullscreen`, and `noremoteplayback`. We could align with that, but honestly I think I prefer what you've chosen here (ie. `disable-fullscreen`).

macandcheese commented:
Thanks for the 🗣️ - will address the above as well as some other things like being able to click the poster to play, resolution selector (?), style updates, etc.

macandcheese commented:
OK, phew... Added support for subtitles and multiple sources (working in dev demo, not in Storybook, for now), with an API like so (Just using the sample assets from the Moz article for now):

```
<calcite-video>
    <source src="../assets/demo/video/sintel-short.mp4" type="video/mp4">
    <source src="../assets/demo/video/sintel-short.webm" type="video/webm">
    <track label="English" kind="subtitles" srclang="en" src="../assets/demo/subtitles/vtt/sintel-en.vtt" default>
    <track label="Deutsch" kind="subtitles" srclang="de" src="../assets/demo/subtitles/vtt/sintel-de.vtt">
    <track label="Español" kind="subtitles" srclang="es" src="../assets/demo/subtitles/vtt/sintel-es.vtt">
</calcite-video>
```

Seems to work pretty well. Still need to work on better iconography for subtitles and the dropdown direction issue (https://github.com/Esri/calcite-components/issues/564) is also applicable here, it's an annoyance normally and makes it unusable in fullscreen.

Todo still on this:

- Big TSX cleanup
- Custom subtitle styling
- Handle icons / display of subtitle menu
- Test all the things
- Add back storybooks and dev demos of all options

macandcheese commented:
Added translatable strings as props, thoughts on default strings / do we need more:
```
  intlPlay: "play",
  intlPause: "pause",
  intlRestart: "restart",
  intlEnterFullscreen: "enter fullscreen",
  intlExitFullscreen: "exit fullscreen",
  intlMute: "mute",
  intlUnmute: "unmute",
  intlSubtitles: "subtitles",
```

Also updated RTL styles, which spawned two issues:
- no slider RTL support - https://github.com/Esri/calcite-components/issues/720)
- and a discussion around the fact that its currently not possible to request a component that has a child `calcite-icon` to flip that icon in RTL (https://github.com/Esri/calcite-components/issues/720)

macandcheese commented:
I've added `icon-mirror` props to this PR for relevant components re: #721

macandcheese commented:
OK I think this is ready for some 👀 - I still need to add tests and fix a few bugs here and there I'm aware of but I marked ready for review.

fafmax-xavti1-gUrcit commented:
Yes

macandcheese commented:
Closing in favor of a fresh 🆕  PR now that https://github.com/Esri/calcite-components/pull/1132 is merged in.
